ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -333,6 +428,118 @@ impl Module {
     pub(crate) fn dump(&self) {
         eprintln!("{}", self);
     }
+
+    /// Push a slice of extra arguments into the extra arg table.
+    fn push_extra_args(&mut self, ops: &[Operand]) -> ExtraArgsIndex {
+        let idx = self.extra_args.len();
+        self.extra_args.extend_from_slice(ops); // FIXME: this clones.
+        ExtraArgsIndex(u16::try_from(idx).unwrap()) // FIXME: propagate error
+    }
+
+    /// Lookup a type by its index. Panics if an out of bounds index is supplied.
+    fn get_type(&self, tidx: TypeIndex) -> &Type {
+        &self.types[usize::from(tidx.0)]
+    }
+
+    /// Push a new type into the type table and return its index.
+    fn push_type(&mut self, ty: Type) -> TypeIndex {
+        let idx = self.types.len();
+        self.types.push(ty);
+        TypeIndex(u16::try_from(idx).unwrap()) // FIXME: propagate error
+    }
+
+    /// Push a new constant into the constant table and return its index.
+    pub(crate) fn push_const(&mut self, constant: Constant) -> ConstIndex {
+        let idx = self.consts.len();
+        self.consts.push(constant);
+        ConstIndex(u16::try_from(idx).unwrap()) // FIXME: propagate error
+    }
+
+    /// Get the index of a type, inserting it in the type table if necessary.
+    pub fn const_index(&mut self, c: &Constant) -> ConstIndex {
+        // FIXME: can we optimise this?
+        for (idx, tc) in self.consts.iter().enumerate() {
+            if tc == c {
+                // const table hit.
+                return ConstIndex(u16::try_from(idx).unwrap()); // FIXME: propagate error
+            }
+        }
+        // type table miss, we need to insert it.
+        self.push_const(c.clone())
```

Comment:
Are we sure that `clone` is a good idea here? If `Constant` is a big chunk of memory, this will have unpleasant performance and maybe other weird consequences.

Reply:
See e.g. the `Entry` API in hashmaps.

Reply:
As far as I can see `entry()` unconditionally takes ownership of the key you give it.

https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html#method.entry

So if the key is a large struct, can you rely on the compiler eliding copying it into `entry()`'s frame? I don't think so.

If not, then you copy the key even if the entry already exists, right? So if you expect more hits than misses (no idea if that's true), the it would be prudent to take a reference instead?

Reply:
>  So if the key is a large struct, can you rely on the compiler eliding copying it into entry()'s frame?

Nope, you can never rely on that sort of thing I'm afraid.

> So if you expect more hits than misses (no idea if that's true), the it would be prudent to take a reference instead?

My guess is that in many cases you can write the code to know whether the function has been encountered yet or not. Maybe I'm wrong.

Reply:
That being the case, I think it's best to accept a reference, as it gives you the flexibility to pass a reference to something you've already looked up and not incur a copy.

Reply:
Since consts are not shared, this comment also remains outstanding.

The JIT will surely want to make its own constants for constant folding.

Reply:
I remember a conversation about sharing the constants with the AOT IR, but allowing the JIT IR to add constants to its own table.

Do we still want to try that?

The other thing we could do is only share constants with the AOT IR and tackle it later when we come to write optimisation passes.

Reply:
Yeah, but in retrospect we can punt on it for this PR.